### DISCUSSION TITLE
Bridging the gap between simplicity and Type Safety in either Session or Engine

### DISCUSSION THREAD

**Discussion by ManiMozaffar** [sqlalchemy/sqlalchemy#11655] [Ideas]
I'm expanding this issue, trying to explain my usecase: https://github.com/sqlalchemy/sqlalchemy/issues/11654

It took me few months to really learn about session, and what is actually going on under the hood.

I learned about snapshot in sqlalchemy by reading the source code. Nowhere in document in Session it actually mentioned about rolling back and restoring snapshot behavior, not at least that I could find it. it's only on "changelogs".
<img width="1198" alt="Screenshot 2024-07-28 at 14 10 55" src="https://github.com/user-attachments/assets/fc70f033-e225-48fb-9c80-894bf55e0378">

so I understood that any object coming from my queries, get attached to session, and despite setting "expire_on_commit=False" they might still actually expire on rollback, or they might even worse roll backed to previous state in memory according to snapshot.

I don't get the idea of it, it's too implicit for my taste, I simply want to construct a query, run the query, and get typed result back from the cursor. That way, no one ever need to read sqlalchemy documentation to understand what's my code is actually doing, it's simply writing some queries and hitting database and that's it. very explicit, very clear and ease learning curve.

So after investigating, turned out Engine is actually what I want, the most stateless component in database interaction in sqlalchemy, where I can just run my queries with, and get typed result. but the problem is that runtime and typing are not in pair, and I really would like to have a stateless component where i can interact with database easily with a good query complier and parser.
SQLAlchcemy already has 99% of it, what do you think about this?

**Comment by ManiMozaffar** [comment]
One solution that I have thought about it is to use session.expunge, before closing the connection context manager. but that's even worse IMO because the interface is still there. having to worry about that state in Session, is a red alert for me. I think that's also why in documentation it said session is not coroutine safe.

**Comment by CaselIT** [comment]
Hi,

It's not clear what you want. This sounds like you want to expunge the object from the session. Can you clarify?

**Reply by ManiMozaffar** [reply]
Hi, thanks for the quick response. Yes, I want to expunge objects from the session, but my main concern is that I don't want the session to manage or retain any state on its own. Having to call session.expunge_all() after every query adds unnecessary complexity and potential for error.

The automatic behaviors of attaching, expiring, and snapshotting objects in SQLAlchemy are too implicit for my taste. I prefer a more explicit approach to database interactions to make the codebase easier to understand and maintain, especially for new developers joining the project.

I explored using the Engine for this purpose, as it seems to offer a more stateless approach to query execution. However, I encountered challenges with the Engine not constructing objects in the same way as the session does, and there's a discrepancy between runtime behavior and static typing. This makes it harder to ensure type safety and clear data handling.

Is there a recommended way to achieve a more stateless interaction model in SQLAlchemy that still allows for robust typing and object construction? Any advice or guidance on this would be greatly appreciated :)

**Reply by CaselIT** [reply]
No, I don't think there is a way of altering the session to make it expunge the objects by default.
You could use a wrapper that creates a session as needed given a connection, does the operations and closes it afterwards

**Reply by ManiMozaffar** [reply]
>You could use a wrapper that creates a session as needed given a connection, does the operations and closes it afterwards

That's smart. Never though about it. Thanks!

>No, I don't think there is a way of altering the session to make it expunge the objects by default.

But can it not achieve this by making engine aware of return type struct? I'm not talking about "attaching object" and object mutation resulting in queries and etc. These are "ORM-like behavior". but just specifying a return struct for a query, should still be part of engine right?

**Reply by CaselIT** [reply]
It's a lot more complex that you make it sound. What about all the other orm fearures, like relationship and similar? 

If something like that were to be implemented it would likely be part of or session, telling it to not track the objects. 

An alternative could be to make it public the ability of specifying a "row constructor", but making non trivial cases work is anything but simple

**Reply by gordthompson** [reply]
re: "making engine aware of return type struct" - There's this:

```python
with engine.connect() as conn:
    rows_as_dict = conn.execute(select(User)).mappings().all()
    print(rows_as_dict)
    """
[{'id': 1, 'name': 'Gord'}, {'id': 2, 'name': 'Barb'}]
    """
```

Is that what you mean?

**Reply by ManiMozaffar** [reply]
@gordthompson Not exactly.
I believe you can achieve the same result as calling Row._as_dict(), would return you the dict response. but sometimes when you join two/three tables, sqla engine annotate fields to different name (like id_anno1) so it can differentiate between a User.id and Author.id 
What I mean is that scalars/execute, could have constructed the actual corresponding dataclass, but i'm not aware of its complexity.


> If something like that were to be implemented it would likely be part of or session, telling it to not track the objects.

I think that is a good idea @CaselIT 

> An alternative could be to make it public the ability of specifying a "row constructor", but making non trivial cases work is anything but simple

That is also a great solution. I think the "orm relationship" is also called implicit relationship right? so how can you be both implicit and explicit at same time haha. Personally I just join tables in my queries, like normal SQL. I never used any implicit ORM relationship. I think that's the main key here, then you'd allow developers to decide about being implicit or explicit.

**Comment by CaselIT** [comment]
> I learned about snapshot in sqlalchemy by reading the source code.

I think what you refer to is documented as transactions: https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#unitofwork-transaction

**Reply by ManiMozaffar** [reply]
> > I learned about snapshot in sqlalchemy by reading the source code.
> 
> I think what you refer to is documented as transactions: https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#unitofwork-transaction

Maybe, but not very explicitly with examples. What i'm talking about is this 2 line of code:
https://github.com/sqlalchemy/sqlalchemy/blob/156fef61135a55c6ad17765b64155801f1dbea66/lib/sqlalchemy/orm/session.py#L1079
https://github.com/sqlalchemy/sqlalchemy/blob/156fef61135a55c6ad17765b64155801f1dbea66/lib/sqlalchemy/orm/session.py#L1101
I did not know that sqlalchemy tracking object also restore object to previous state, before the transaction is rolled back.

**Reply by CaselIT** [reply]
It doesn't. It expires them.

Meaning that if you load an object, change an attribute value, then call rollback, you don't see the updated value in the object.

Seems a reasonable automation, considering that orm objects are asdumed to represent database rows
